# Exploring eBay Car Sales Data
## Executive Summary

### The structure of used auto market in Gemany in 2016 as follows:
Upper middle class (Audi, Mercedes_bez, BMW): 29.29%
Middle class (Volkswagen and Seat): 22.92%
Lower middle class (Ford, Peugeot, Opel, Fiat, Renault): 17.24%
Other brands: 30.55%

In the second hand car market, the mileage of more luxury cars usually greater than that of lower segments.
Both reputation of brand and mileage affect the car price, but the effect of brand's reputation is much more significant than that of the number km recorded.

### Most popular model of each brands available in the market:

- 'volkswagen': ('golf', 3707),
- 'bmw': ('3er', 2615),
- 'opel': ('corsa', 1592),
- 'mercedes_benz': ('c_klasse', 1136),
- 'audi': ('a4', 1231),
- 'ford': ('focus', 762),
- 'renault': ('twingo', 615),
- 'peugeot': ('2_reihe', 600),
- 'fiat': ('punto', 415),
- 'seat': ('ibiza', 328)

However, Volkswagen domains this auto market, so 3 out of 10 most common models are all marked Volkswagen. In the same brand, the car with damages usually costs 3 times more than its counterparts without damages.

The avarage gap price between Automatic and Manual gear box is $6256.

## Data File Explanation
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset isn't available on Kaggle anymore, but you can find it [at the link here](https://data.world/data-society/used-cars-data).

### Modifications to file
- This is sample of 50,000 data points from the full dataset
- data file was intentionally "dirtied" to practice data cleanup skills in python and panda

### Data dictionary definitions
- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

## Project Goals
- Clean data
- analyze data using panda library

## Import data into notebook

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observations
The dataset contains 20 columns, most of which are strings. Some columns contain null values:
- `vehicleType`
- `gearbox`
- `model`
- `fuelType`
- `notRepairedDamage`

Additionally, we noted that the column names are using camelcase (preferredColumn) instead of the preferred snakecase (preferred_column). Finally, we also noted some 0 values for some rows in PowerPS, which may lead to issues further on trying to look for averages in that column.

To begin, we will convert the columnnames from camelcase to snakecase, as well as renaming some the columns to be more appropriately descriptive.


In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
#Copying the dataframe:
copy_autos = autos.copy()

#W1: One way to change the names of columns:
copy_autos.rename({'dateCrawled': 'date_crawled','offerType' : 'offer_type',
       'vehicleType':'vehicle_type', 'yearOfRegistration':'registration_year', 'gearbox':'gear_box', 'powerPS':'power_PS',
       'monthOfRegistration':'registration_month', 'fuelType':'fuel_type', 'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}, inplace = 1, axis = 1)
copy_autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [5]:
autos = copy_autos.copy()
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [ ]:
#Copying the dataframe:
autos_copy_2 = autos.copy()
#W2: another way to change names of columns:
autos_copy_2.columns = ['date_crawled','name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gear_box', 'power_PS', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos_copy_2.head()

In [ ]:
#W3: alternative way to change names of columns: Series.map(mapping_dict)
# Note: if a value from our series doesn't exist as a key in our dictionary, it will convert that value to NaN
mapping_dict = {'dateCrawled':'date_crawled',
                'name':'name',
                'seller':'seller',
                'price':'price',
                'abtest':'abtest',
                'model':'model',
                'odometer':'odometer',
                'brand':'brand',
                'offerType':'offer_type',
                'vehicleType':'vehicle_type',
                'yearOfRegistration':'registration_year',
                'gearbox':'gear_box',
                'powerPS':'power_PS', 
                'monthOfRegistration':'registration_month',
                'fuelType':'fuel_type',
                'notRepairedDamage':'unrepaired_damage',
                'dateCreated':'ad_created', 
                'nrOfPictures':'nr_of_pictures', 
                'postalCode':'postal_code',
                'lastSeen':'last_seen'}
autos.columns=autos.columns.map(mapping_dict)
autos.head()

## Basic Data Exploration
Initially we will look for
- Text columns where all or almost all values are the same. These can be dropped as they are not useful for analysis
- numeric data stored as text which can be cleaned and converted.

In [6]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Basic Exploration Observations
As noted earlier, power_ps has some 0 values, but registration_month also has 0's. Price we should review further as well, as the top value is "$0".

`price` and `odometer` are numeric values stored as text. we will need to remove non-numeric characters and convert each column to a numeric dtype.

Some candidates for dropping:
- `registration_month`

In [7]:
# Cleaning and adjusting price and odometer columns
autos["price"] = autos["price"].str.replace('$','').str.replace(',','').astype(int)
autos["odometer"] = autos["odometer"].str.replace('km','').str.replace(',','').astype(int)
autos.rename({"price":"price_usd", "odometer": "odometer_km"}, axis = 1, inplace = True)
print(autos['price_usd'].head())
print('\n')
print(autos['odometer_km'].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_usd, dtype: int32


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32


In [8]:
# Mapping german to english
mapping_dic ={'manuell': 'manual',
              'automatik':'automatic',
              'lpg':'lpg',
              'benzin': 'petrol',
              'diesel':'diesel',
              'cng': 'cng',
              'hybrid':'hybrid',
              'elektro':'electric',
              'andere':'others',
              'nein':'no',
              'ja':'yes'
              } #set up a dictionary to translate german words to english
cols_translated = ['gear_box','fuel_type','unrepaired_damage'] #list containing column names targeted
for c in cols_translated: # running loop off of new list
    autos[c]=autos[c].map(mapping_dic) #use current iteration to update values in autos based on mapping_dic
    
print(autos['unrepaired_damage'].unique())

['no' nan 'yes']


In [9]:
autos['nr_of_pictures'].value_counts() # all rows are 0 values, and is meaningless for our analysis

0    50000
Name: nr_of_pictures, dtype: int64

In [10]:
autos['seller'].value_counts() # only 1 seller is 'gewerblich', so this is meaningless

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos['offer_type'].value_counts() # same here - with one 'gesuch', this is meaningless

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
# dropping the tables that are of no use for our analysis
autos.drop(['nr_of_pictures','seller','offer_type'], axis = 1, inplace = True)

## Exploration Continued - `price_usd` and `odometer_km`
From the last step, we learned that there are a number of text columns where almost all of the values are the same (`seller` and `offer_type`). We also converted the `price` and `odometer` columns to numeric types and renamed `price` to `price_usd` `odometer` to `odometer_km`.

We will continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns:

In [13]:
# price_usd column:
print(autos['price_usd'].describe())
print('\n')
print('Number of unique values:',autos['price_usd'].unique().shape[0])
print('\n')
print('Top 10 cheapest deals before: ',autos['price_usd'].value_counts().sort_index(ascending = True).head(15))
print('\n')
print('Top 10 most expensive deals before: ',autos['price_usd'].value_counts().sort_index(ascending = False).head(15))

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64


Number of unique values: 2357


Top 10 cheapest deals before:  0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price_usd, dtype: int64


Top 10 most expensive deals before:  99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price_usd, dtype: int64


In our samples, the prices of all the deal were provided. There are only 2357 unique values, this seems to be the result of people's tendency to round prices on the site. The lowest value is $0.00 (with 1421 deals) which is about 2.8 percent of total deals so we can remove them; and the highest value is 99999999. Let's have a closer look in this row:

In [14]:
autos[autos['price_usd'] == 99999999]

date_crawled                         name  price_usd   abtest  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges   99999999  control   

      vehicle_type  registration_year   gear_box  power_PS     model  \
39705    limousine               1999  automatic       224  s_klasse   

       odometer_km  registration_month fuel_type          brand  \
39705       150000                   9    petrol  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
39705               NaN  2016-03-22 00:00:00        73525  2016-04-06 05:15:30

For the kind of car this row describes, the price appears to be incorrect, so we will drop this row.

Also, there appears to be a big difference in pricing from 350,000 USD to 999,999USD. We will check to confirm the number of transactions between 1 dollar and 350 thousand usd:

In [15]:
per_1_350000 = (autos[autos['price_usd'].between(1,350000)].shape[0]/50000)*100
print(round(per_1_350000,2),'%')

97.13 %


In [20]:
# drop rows with price == 0 usd:
autos.drop(autos.index[autos['price_usd'] == 0].tolist(), axis = 1, inplace = True)
#drop rows with price greater than 1 and less than 350000 USD:
list_index_350000 = autos.index[autos['price_usd'] > 350000].tolist() #create index matching conditions
autos.drop(list_index_350000, axis = 0, inplace = True) #dropping the rows
autos['price_usd'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

In [21]:
# 'odometer_km' column
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

## Exploring date columns
There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [22]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price_usd           48565 non-null  int32 
 3   abtest              48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gear_box            46222 non-null  object
 7   power_PS            48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int32 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [23]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use `Series.str[:10]`:

In [24]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


In [31]:
# date_crawled column
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = False)

2016-04-07    0.001400
2016-04-06    0.003171
2016-04-05    0.013096
2016-04-04    0.036487
2016-04-03    0.038608
2016-04-02    0.035478
2016-04-01    0.033687
2016-03-31    0.031834
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-28    0.034860
2016-03-27    0.031092
2016-03-26    0.032204
2016-03-25    0.031607
2016-03-24    0.029342
2016-03-23    0.032225
2016-03-22    0.032987
2016-03-21    0.037373
2016-03-20    0.037887
2016-03-19    0.034778
2016-03-18    0.012911
2016-03-17    0.031628
2016-03-16    0.029610
2016-03-15    0.034284
2016-03-14    0.036549
2016-03-13    0.015670
2016-03-12    0.036920
2016-03-11    0.032575
2016-03-10    0.032184
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-07    0.036014
2016-03-06    0.014043
2016-03-05    0.025327
Name: date_crawled, dtype: float64

Crawling of the site occured continuously between 03/05/2016 and 04/07/2016. The distribution is pretty uniform, roughly 3% per day in terms of frequency.

In [30]:
# ad_created column
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = False)

2016-04-07    0.001256
2016-04-06    0.003253
2016-04-05    0.011819
2016-04-04    0.036858
2016-04-03    0.038855
                ...   
2015-12-05    0.000021
2015-11-10    0.000021
2015-09-09    0.000021
2015-08-10    0.000021
2015-06-11    0.000021
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates, from 06/11/2015 to 04/07/2016. There are far less ads created in the beginning, which makes sense if the site was just starting up and not as popular.

In [32]:
# last_seen column
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = False)

2016-04-07    0.131947
2016-04-06    0.221806
2016-04-05    0.124761
2016-04-04    0.024483
2016-04-03    0.025203
2016-04-02    0.024915
2016-04-01    0.022794
2016-03-31    0.023783
2016-03-30    0.024771
2016-03-29    0.022341
2016-03-28    0.020859
2016-03-27    0.015649
2016-03-26    0.016802
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-23    0.018532
2016-03-22    0.021373
2016-03-21    0.020632
2016-03-20    0.020653
2016-03-19    0.015834
2016-03-18    0.007351
2016-03-17    0.028086
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-14    0.012602
2016-03-13    0.008895
2016-03-12    0.023783
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-08    0.007413
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

The date ranges makes sense, as it should mimic the dates with crawling. We can infer that a value in this column would identify the date that a listing was removed, possibly a sale or the posting to sell was withdrawn.

There are a disproportionate amount of values recorded in the last 3 days. Instead of a massive sale of cars in the period, it is likely more to do with the end of the data crawling than anything else.

In [34]:
print(autos['registration_year'].describe())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


We have some odd values in our registration years, notably '0000' and '9999'. We should review this further:

In [35]:
# registration_year column
autos['registration_year'].value_counts(normalize = True, dropna = False).sort_index(ascending = False)

9999    0.000062
9000    0.000021
8888    0.000021
6200    0.000021
5911    0.000021
          ...   
1910    0.000103
1800    0.000041
1111    0.000021
1001    0.000021
1000    0.000021
Name: registration_year, Length: 95, dtype: float64

## Dealing with incorrect registration years

We have two kinds of values to correct in registration years. As this is data containing ads from up 2016, we shouldn't have any years recorded after that. Also, we shouldn't expect any cars earlier than 1885 at the bare minimum, although we should expect some antiques in our listings.

First, let us review the count of how many car listings have registration years before 1900 or after 2016:

In [43]:
bool_array = (autos['registration_year']< 1900)|(autos['registration_year']> 2016)
percent = (autos[bool_array]).shape[0]/autos.shape[0]*100
print(round(percent,2),'%')
# alternative code to count the number of listings with cars that fall outside the 1900 - 2016 interval: 
#(~autos['registration_year'].between(1900,2016)).sum()

3.88 %


Knowing this is less than 4% of our listings, we are going to drop these rows


In [44]:
autos= autos[autos['registration_year'].between(1910,2016)] #only includes reg years between 1910 and 2016
print(autos['registration_year'].describe()) #checking distribution
autos['registration_year'].value_counts(normalize = True, dropna = False).sort_values()[-15:]

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64


2010    0.034040
2011    0.034768
1997    0.041794
2009    0.044665
2008    0.047450
2007    0.048778
1998    0.050620
2002    0.053255
2001    0.056468
2006    0.057197
2003    0.057818
2004    0.057904
1999    0.062060
2005    0.062895
2000    0.067608
Name: registration_year, dtype: float64

The distribution definitely looks a lot better after the cleanup. We can note there is a significant amount of car registrations between 1997-2008 in our data, over 60% consisting in that timeframe.

## Aggregation of data - brand

In [47]:
autos['brand'].value_counts(normalize = True, dropna = False).sort_values(ascending = False)[:10]

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64

The initial observation is that, as expected, German brands are dominating in the market, accounting for nearly 55% of the overall listings. This is followed by American brands like Opel and Ford which takes 17% market share, and tailing is French brands like Renault and Peugeot (~8%).

Volkswagon is the most popular brand (21%), double the market share for BMW. With over 80% of the listings accounted for in the top 10 brands, we will just focus our analysis on these.

In [48]:
sorted_brands = autos['brand'].value_counts(normalize = True, dropna = False).sort_values(ascending = False)[:10]

In [49]:
brands = sorted_brands.index #creating an index of our brands
print(brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [51]:
brand_price = {} #empty dictionary
for b in brands: #loop over selected brands
    mean_price = autos.loc[autos['brand']==b,'price_usd'].mean() #assign mean price by brand to mean_price
    brand_price[b] = round(mean_price,2) #assign to brand_price dictionary key with [b] and value to mean_price
print(brand_price)

{'volkswagen': 5402.41, 'bmw': 8332.82, 'opel': 2975.24, 'mercedes_benz': 8628.45, 'audi': 9336.69, 'ford': 3749.47, 'renault': 2474.86, 'peugeot': 3094.02, 'fiat': 2813.75, 'seat': 4397.23}


In [52]:
sorted_price = sorted(brand_price.items(), key=lambda x: x[1], reverse = True)    
print(sorted_price)

[('audi', 9336.69), ('mercedes_benz', 8628.45), ('bmw', 8332.82), ('volkswagen', 5402.41), ('seat', 4397.23), ('ford', 3749.47), ('peugeot', 3094.02), ('opel', 2975.24), ('fiat', 2813.75), ('renault', 2474.86)]


While we know Volkswagen to be the most popular brand by the amount of listings, it's average price is pretty average compared to the other brands. The American and French brands are on average less expensive, averaging between two to four thousand usd. The most expensive brands by far are audi, mercedes benz and BMW, averaging 8 to 9 thousand in usd.

Based on this data, the second-hand car market in Germany is dominated by used middle to high end brands like VW, BMW and Mercedes Benz.

## Correlating car price with mileage

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we will use the following for the next excercise:
- `pandas series constructor`
- `pandas dataframe constructor`

In [55]:
bmp_series = pd.Series(brand_price) #converting our dictionary from before to a series
print(bmp_series) # the key from the dict becomes index, the values as a new column w/ no header

volkswagen       5402.41
bmw              8332.82
opel             2975.24
mercedes_benz    8628.45
audi             9336.69
ford             3749.47
renault          2474.86
peugeot          3094.02
fiat             2813.75
seat             4397.23
dtype: float64


In [56]:
df = pd.DataFrame(bmp_series, columns=['mean_price']) #converts our series into a new dataframe
df #note the need to name our columns

mean_price
volkswagen        5402.41
bmw               8332.82
opel              2975.24
mercedes_benz     8628.45
audi              9336.69
ford              3749.47
renault           2474.86
peugeot           3094.02
fiat              2813.75
seat              4397.23

Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.

In [58]:
# storing the mean prices and mileage into their own dictionaries
brand_mean_prices = {} #empty dictionary
for b in brands: #loop over selected brands
    mean_prices = autos.loc[autos['brand']==b,'price_usd'].mean() #assign mean price by brand to mean_price
    brand_mean_prices[b] = round(mean_prices,2) #assign to brand_price dictionary key with [b] and value to mean_price
print(brand_mean_prices)

brand_mean_mileage = {}
for b in brands:
    mean_mileage =autos.loc[autos['brand']==b, 'odometer_km'].mean()
    brand_mean_mileage[b] = round(mean_mileage,2)
print(brand_mean_mileage)

{'volkswagen': 5402.41, 'bmw': 8332.82, 'opel': 2975.24, 'mercedes_benz': 8628.45, 'audi': 9336.69, 'ford': 3749.47, 'renault': 2474.86, 'peugeot': 3094.02, 'fiat': 2813.75, 'seat': 4397.23}
{'volkswagen': 128707.16, 'bmw': 132572.51, 'opel': 129310.04, 'mercedes_benz': 130788.36, 'audi': 129157.39, 'ford': 124266.01, 'renault': 128071.33, 'peugeot': 127153.63, 'fiat': 117121.97, 'seat': 121131.3}


In [60]:
# turning our new dictionaries into series:
bmp_series = pd.Series(brand_mean_prices) #converting our dictionary from before to a series
bmm_series = pd.Series(brand_mean_mileage)

mean_dataframe = pd.DataFrame(bmp_series, columns = ['mean_price']) #converts our series into a new dataframe
mean_dataframe #note the need to name our columns


mean_price
volkswagen        5402.41
bmw               8332.82
opel              2975.24
mercedes_benz     8628.45
audi              9336.69
ford              3749.47
renault           2474.86
peugeot           3094.02
fiat              2813.75
seat              4397.23

In [61]:
mean_dataframe['mean_mileage'] = bmm_series

In [62]:
mean_dataframe

mean_price  mean_mileage
volkswagen        5402.41     128707.16
bmw               8332.82     132572.51
opel              2975.24     129310.04
mercedes_benz     8628.45     130788.36
audi              9336.69     129157.39
ford              3749.47     124266.01
renault           2474.86     128071.33
peugeot           3094.02     127153.63
fiat              2813.75     117121.97
seat              4397.23     121131.30

There is no real difference in mileage between the brands at this level. There is possibly some correlation when looking at the higher-end brands and their respective pricing/mileage but we will need more analysis to confirm this.

To get to that, we will segment the higher end brands into three segments and compare pricing/mileage there:

In [63]:
upper_middle_class = autos[(autos['brand'] == 'bmw')|(autos['brand'] == 'audi')|(autos['brand'] == 'mercedes_benz')]
upper_middle_class['brand'].value_counts()

bmw              5137
mercedes_benz    4503
audi             4041
Name: brand, dtype: int64

In [64]:
# splitting odomoter_km into 3 groups
mileage_groups = upper_middle_class["odometer_km"].value_counts(bins = 3).sort_index()
mileage_groups

(4854.999, 53333.333]        986
(53333.333, 101666.667]     1689
(101666.667, 150000.0]     11006
Name: odometer_km, dtype: int64

In [65]:
price1 = upper_middle_class.loc[upper_middle_class["odometer_km"] <= 53333.333,'price_usd'].mean() #group 1
print('(4854.999, 53333.333] __ Average price: $', round(price1,2))
price2 = upper_middle_class.loc[upper_middle_class["odometer_km"].between(53333.333, 101666.667),'price_usd'].mean() #group 2
print('(53333.333, 101666.667] __ Average price: $', round(price2,2))
price3 = upper_middle_class.loc[upper_middle_class["odometer_km"] >= 101666.667,'price_usd'].mean() #group 3
print('(101666.667, 150000.0] __ Average price: $', round(price3,2))

(4854.999, 53333.333] __ Average price: $ 23037.13
(53333.333, 101666.667] __ Average price: $ 16511.87
(101666.667, 150000.0] __ Average price: $ 6249.86


It is clear that in the same segment and as expected, the cars with higher mileage will cost less compared with the counterparts with less mileage. However, the reputation of the brand affects more to the price of the listing than the mileage the car recorded - we saw that in our original analysis. In the secondhand market, the luxury cars' mileages are usually higher than the lower middle class cars' ones.

## Find the most common brand / model combinations
We would like to have more specific information of the hotest models consumed of our top 10 brands.

In [66]:
# creating model dictionary
top_brand_model = {}
def hot_model(name):
    model=autos[autos['brand']==name]['model'].value_counts()
    model_name = model.index[0]
    model_size = model[0]
    return model_name,model_size
for model in brand_price:
    top_brand_model[model] = hot_model(model)
print('Hottest models of top 10 brands: ')
top_brand_model

Hottest models of top 10 brands: 


{'volkswagen': ('golf', 3707),
 'bmw': ('3er', 2615),
 'opel': ('corsa', 1592),
 'mercedes_benz': ('c_klasse', 1136),
 'audi': ('a4', 1231),
 'ford': ('focus', 762),
 'renault': ('twingo', 615),
 'peugeot': ('2_reihe', 600),
 'fiat': ('punto', 415),
 'seat': ('ibiza', 328)}

## What are the hottest selling models, and which brands?

This time, instead of using aggregation, we will try to use `groupby()`.

A groupby operation involves:
- splitting the object
- applying a function
- combining the results



In [78]:
autos.groupby('brand')['model'].value_counts().sort_values(ascending=False).head(10)
# another way to call a series from a DataFrame: using dot notation so our alternative code:
#autos.groupby('brand').model.value_counts().sort_values(ascending = False).head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
Name: model, dtype: int64

3 of the top 10 models are made by volkswagen, with the top one being the golf at 3707 listings. This is consistent with our understanding that volkswagen takes up the largest market share at around 21%.

## How much cheaper are damaged cars than those without damage?

In [79]:
# practicing more with groupby()
autos.groupby(['unrepaired_damage']).price_usd.mean().sort_values(ascending=False)

unrepaired_damage
no     7164.033103
yes    2241.146035
Name: price_usd, dtype: float64

In [80]:
upper_middle_class.groupby('unrepaired_damage').brand.value_counts()

unrepaired_damage  brand        
no                 bmw              3948
                   mercedes_benz    3465
                   audi             3064
yes                bmw               403
                   mercedes_benz     343
                   audi              333
Name: brand, dtype: int64

In [81]:
upper_middle_class.groupby(['unrepaired_damage','brand']).price_usd.mean()

unrepaired_damage  brand        
no                 audi             10914.959856
                   bmw               9437.709980
                   mercedes_benz     9798.396537
yes                audi              3324.684685
                   bmw               3512.637717
                   mercedes_benz     3921.819242
Name: price_usd, dtype: float64

In [83]:
middle_class = autos[(autos['brand']=='volkswagen')|(autos['brand']=='Seat')]
middle_class.groupby(['unrepaired_damage', 'brand']).price_usd.mean()

unrepaired_damage  brand     
no                 volkswagen    6469.407759
yes                volkswagen    2179.405660
Name: price_usd, dtype: float64

we can conclude that in the same brand, the cars without damages are usually 3 times more expensive than the damaged cars.

## Does gearbox affect price?

In [85]:
autos.groupby('gear_box').price_usd.mean()

gear_box
automatic    10972.718547
manual        4716.709175
Name: price_usd, dtype: float64

In [87]:
autos.groupby('gear_box').price_usd.mean()['automatic'] - autos.groupby('gear_box').price_usd.mean()['manual']

6256.009372369583

The average price difference between an automatic and manual vehicle is about 6200 usd.